File: fs/src/main/java/io/zbox/fs/VersionReader.java
package io.zbox.fs;

import java.nio.ByteBuffer;

public class VersionReader extends RustObject {

    private static int rustObjId = 104;

    // read buffer
    private int READ_BUF_CAP = 16 * 1024;
    private ByteBuffer readBuf = ByteBuffer.allocateDirect(READ_BUF_CAP);

    private VersionReader() {
    }

    public long read(ByteBuffer dst) throws ZboxException {
        checkNullParam(dst);

        if (dst.isDirect()) {
            long ret = this.jniRead(dst.slice());
            dst.position(dst.position() + (int) ret);
            return ret;
        }

        ByteBuffer cloned = ByteBuffer.allocateDirect(dst.remaining());
        long ret = this.jniRead(cloned);
        cloned.limit((int) ret);
        dst.put(cloned);
        return ret;
    }

    public int read(byte[] dst, int off, int len) throws ZboxException {
        ByteBuffer buf;

        if (len > READ_BUF_CAP) {
            buf = ByteBuffer.allocateDirect(len);
        } else {
            buf = this.readBuf.duplicate();
            buf.clear();
            buf.limit(len);
            buf = buf.slice();
        }

        int ret = (int) this.jniRead(buf);
        buf.limit(ret);
        buf.get(dst, off, len > ret ? ret : len);

        return ret;
    }

    public int read(byte[] dst) throws ZboxException {
        return this.read(dst, 0, dst.length);
    }

    public ByteBuffer readAll() throws ZboxException {
        ByteBuffer ret = this.jniReadAll();
        ret.position(ret.limit());
        return ret;
    }

    public long seek(long offset, SeekFrom whence) throws ZboxException {
        checkNullParam(whence);
        return this.jniSeek(offset, whence.getValue());
    }

    // jni methods
    private native long jniRead(ByteBuffer dst) throws ZboxException;

    private native ByteBuffer jniReadAll() throws ZboxException;

    private native long jniSeek(long offset, int whence) throws ZboxException;
}

File: zbox/src/version_reader.rs
use std::io::{Read, Seek, SeekFrom};

use jni::objects::{JByteBuffer, JObject, JValue};
use jni::sys::{jint, jlong};
use jni::JNIEnv;

use zbox::{Error, VersionReader};

use super::{throw, to_seek_from, RUST_OBJ_FIELD};

#[no_mangle]
pub extern "system" fn Java_io_zbox_zboxfs_VersionReader_jniRead(
    env: JNIEnv,
    obj: JObject,
    dst: JByteBuffer,
) -> jlong {
    let mut rdr = env
        .get_rust_field::<&str, VersionReader>(obj, RUST_OBJ_FIELD)
        .unwrap();
    let dst = env.get_direct_buffer_address(dst).unwrap();
    match rdr.read(dst) {
        Ok(read) => read as i64,
        Err(err) => {
            throw(&env, Error::from(err));
            0
        }
    }
}

#[no_mangle]
pub extern "system" fn Java_io_zbox_zboxfs_VersionReader_jniReadAll<'a>(
    env: JNIEnv<'a>,
    obj: JObject,
) -> JByteBuffer<'a> {
    let mut rdr = env
        .get_rust_field::<&str, VersionReader>(obj, RUST_OBJ_FIELD)
        .unwrap();

    // get file content length
    let content_len = match rdr.version() {
        Ok(ver) => ver.content_len(),
        Err(err) => {
            let ret = JObject::null();
            throw(&env, err);
            return JByteBuffer::from(ret);
        }
    };

    // get current position of file
    let curr_pos = match rdr.seek(SeekFrom::Current(0)) {
        Ok(pos) => pos as usize,
        Err(err) => {
            let ret = JObject::null();
            throw(&env, Error::from(err));
            return JByteBuffer::from(ret);
        }
    };

    // calculate direct buffer length needed
    let len = if content_len > curr_pos {
        content_len - curr_pos
    } else {
        0
    };

    // allocate a direct byte buffer on Java side, this is to let JVM to handle
    // buffer release
    let buf_obj = env
        .call_static_method(
            "java/nio/ByteBuffer",
            "allocateDirect",
            "(I)Ljava/nio/ByteBuffer;",
            &[JValue::from(len as i32)],
        )
        .unwrap()
        .l()
        .unwrap();
    let buf = JByteBuffer::from(buf_obj);
    let dst = env.get_direct_buffer_address(buf).unwrap();

    // read file content to direct byte buffer
    let mut offset = 0;
    loop {
        match rdr.read(&mut dst[offset..]) {
            Ok(read) => {
                if read == 0 {
                    break;
                }
                offset += read;
            }
            Err(err) => {
                let ret = JObject::null();
                throw(&env, Error::from(err));
                return JByteBuffer::from(ret);
            }
        }
    }

    // set direct buffer limit
    env.call_method(
        *buf,
        "limit",
        "(I)Ljava/nio/Buffer;",
        &[JValue::from(offset as i32)],
    )
    .unwrap();

    buf
}

#[no_mangle]
pub extern "system" fn Java_io_zbox_zboxfs_VersionReader_jniSeek(
    env: JNIEnv,
    obj: JObject,
    offset: jlong,
    whence: jint,
) -> jlong {
    let mut rdr = env
        .get_rust_field::<&str, VersionReader>(obj, RUST_OBJ_FIELD)
        .unwrap();
    let whence = to_seek_from(offset, whence);
    match rdr.seek(whence) {
        Ok(pos) => pos as i64,
        Err(err) => {
            throw(&env, Error::from(err));
            0
        }
    }
}
